### PR TITLE
Allow window title bars to be removed

### LINKED ISSUE #217: Allow window title bars to be removed
Provide an option for title bars to be removed (based on @lauverityhill's comment at https://github.com/workspacer/workspacer/issues/212#issuecomment-869144310)

window with no title bar 

https://github.com/workspacer/workspacer/issues/212#issuecomment-869242758 by @system32uwu 
> This can probably be done using the Win32 API, win3wm does this here

https://github.com/workspacer/workspacer/issues/212#issuecomment-869272380 by @dalyIsaac 
> nog has something similar - https://github.com/TimUntersberger/nog/blob/master/assets/default_config.lua
nog also has has_custom_titlebar which allows special handling for some executables.

https://github.com/workspacer/workspacer/issues/212#issuecomment-869289011 by @schiltz3 
> I played around with removing the Title Bar, and if someone does implement this. I would heavily recommend doing it in a way that supports special handling for some executables because messing with the GWL_STYLE is very inconsistent.

### PR DESCRIPTION
**Pull request by dalyIsaac**
Resolves #217 

This adds the ability to hide window title bars, as a new plugin. This only applies to windows created _after_ workspacer starts up.

**Changes:**

- Added `workspacer.TitleBar` project.
- Changed `Win32.WS` and `Win32.WS_EX` to be `long` instead of `ulong`.
  I did this because Win32 functions like [`SetWindowLongPtr`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowlongptrw) specify [`LONG_PTR`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/bbcfb0af-8349-4e98-ad26-957e1363f714) instead of [`ULONG_PTR`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/21eec394-630d-49ed-8b4a-ab74a1614611) (though I'm not sure it matters much).
- Added methods `SetWindowStylePtr` and `SetWindowStyleExLongPtr` to set window styles.
- Add the ability to set/remove `WS_CAPTION` and `WS_THICKFRAME` (see [docs.microsoft.com: Window Styles](https://docs.microsoft.com/en-us/windows/win32/winmsg/window-styles)).
- Added documentation to `config.md`

**Available window styles:**

image

(NOTE: screenshot taken on Windows 11, hence the rounded borders)

### CODE REVIEW

**Review by sitiom — APPROVED** [review]

**Comment by dalyIsaac** [comment]
@sitiom thanks for fixing those build issues - they were driving me insane!

**Comment by lauverityhill** [comment]
Thank you for this!

**Comment by josteink** [comment]
Seems like this PR currently has some conflicts with git master. 

Fix it up and prep for another review? 🙂

**Comment by josteink** [comment]
This PR separates test-projects from normal code by putting it in a different, "less important" folder. 

As someone who believes in rigid test-automation, I think this is a wrong move. 

When I introduced xUnit tests to workspacer (that was me), I intentionally placed all the test-projects alongside normal production code, because the test-suite is required to ship. 

IOW: The tests are production code too and should be treated as such. 

And the best way to keep them like that is to keep them close to the code they're testing. 

As such, moving them *away* from the code under test is a dramatically opposite move to what I consider good form, that I will reject this PR until that is reverted. 

Also: A single PR really should focus on changing only one thing. 

Right now this PR conflates several changes into one branch, causing merge-conflicts and preventing the good bits in there (there are!) from being merged. 

Please try to be a bit more disciplined when making feature-branches in the future. Keeping them small is a *good thing*.

**Review by josteink — CHANGES_REQUESTED** [review]
Put tests back in normal source-folder.

Having tests separate from code being tested makes finding tests and writing tests harder, and is the first step towards code-rot in the test-suite.

**Comment by sitiom** [comment]
> This PR separates test-projects from normal code by putting it in a different, "less important" folder.

It's not a less important folder, it keeps things organized and that's how most projects actually do it.

> When I introduced xUnit tests to workspacer (that was me), I intentionally placed all the test-projects alongside normal production code, because the test-suite is required to ship.
>
> IOW: The tests are production code too and should be treated as such.

Why would production code be dependent on a test suite? Why bloat assemblies with unit tests? 

Even currently, the unit tests are not shipped with the production code and the change only separates the test projects with no actual change with the output anyway.

> Also: A single PR really should focus on changing only one thing.

Although I agree it to be in a separate PR, sure.

File: src/workspacer.Native/Native/Win32/Win32.Long.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ namespace workspacer
     public static partial class Win32
     {
         [Flags]
-        public enum WS : ulong
+        public enum WS : long
```
Comment (josteink) [inline src/workspacer.Native/Native/Win32/Win32.Long.cs:13]: Any reason this needed to be change from `ulong` to regular `long`?

File: src/workspacer.Native/Native/Win32/Win32.Long.cs [hunk 2]
```diff
@@ -37,7 +37,7 @@ public enum WS : ulong
         }
 
         [Flags]
-        public enum WS_EX : ulong
+        public enum WS_EX : long
```
Comment (josteink) [inline src/workspacer.Native/Native/Win32/Win32.Long.cs:40]: Any reason this needed to be change from `ulong` to regular `long`?

File: src/workspacer.Tests/workspacer.Tests.csproj [hunk 3]
```diff
@@ -15,9 +11,13 @@
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
   </ItemGroup>
+  <ItemGroup>
```
Comment (josteink) [inline src/workspacer.Tests/workspacer.Tests.csproj:14]: The changes to this project file seems to be non-significant and should be reverted to create a smaller diff/create fewer merge-conflicts.

File: src/workspacer/workspacer.csproj [hunk 4]
```diff
@@ -18,26 +18,36 @@
...
     <ProjectReference Include="..\workspacer.Shared\workspacer.Shared.csproj" />
+
+    <!-- Build-time dependencies -->
+    <ProjectReference Include="..\workspacer.Bar\workspacer.Bar.csproj" ReferenceOutputAssembly="false" />
+    <ProjectReference Include="..\workspacer.Gap\workspacer.Gap.csproj" ReferenceOutputAssembly="false" />
+    <ProjectReference Include="..\workspacer.ActionMenu\workspacer.ActionMenu.csproj" ReferenceOutputAssembly="false" />
+    <ProjectReference Include="..\workspacer.FocusIndicator\workspacer.FocusIndicator.csproj" ReferenceOutputAssembly="false" />
+    <ProjectReference Include="..\workspacer.Titlebar\workspacer.Titlebar.csproj" ReferenceOutputAssembly="false" />
+    <ProjectReference Include="..\workspacer.Watcher\workspacer.Watcher.csproj" ReferenceOutputAssembly="false" />
   </ItemGroup>
```
Comment (josteink) [inline src/workspacer/workspacer.csproj:32]: This section looks new, and mostly unrelated to the change at hand. What specifically does it change/improve?

File: src/workspacer.TitleBar/TitleBarPlugin.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        public void AfterConfig(IConfigContext context)
+        {
+            _context = context;
+            _context.Workspaces.WindowAdded += WindowAdded;
+        }
+
+        private void SetStyle(IWindow window, TitleBarStyle settings)
+        {
+            var style = (Win32.WS)Win32.GetWindowLongPtr(window.Handle, Win32.GWL_STYLE);
+            if (settings.ShowTitleBar)
```
Comment (josteink) [inline src/workspacer.TitleBar/TitleBarPlugin.cs:34]: If manipulation of the `style`-variable was extracted to its own method, this code could quite easily be tested through unit-tests.

File: src/workspacer/workspacer.csproj (hunk 4 above)
Comment (sitiom) [reply src/workspacer/workspacer.csproj:32]: It is this commit https://github.com/workspacer/workspacer/pull/246/commits/f290558aa3e25fa9116a65be417b1a9b64699928. It adds build-time dependencies so they are ensured to be built first so the CI succeeds.

File: src/workspacer.TitleBar/TitleBarPlugin.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+            if (settings.ShowSizingBorder)
+            {
+                style |= Win32.WS.WS_THICKFRAME;
+            }
+            else
+            {
+                style &= ~Win32.WS.WS_THICKFRAME;
+            }
+            Win32.SetWindowLongPtr(window.Handle, Win32.GWL_STYLE, (IntPtr)style);
```
Comment (josteink) [inline src/workspacer.TitleBar/TitleBarPlugin.cs:51]: If manipulation of the `style`-variable was extracted to its own method, this code could quite easily be tested through unit-tests.

File: workspacer.sln [hunk 7]
```diff
@@ -26,11 +25,15 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "workspacer.ActionMenu", "sr
...
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "workspacer.Shared.Tests", "src\workspacer.Shared.Tests\workspacer.Shared.Tests.csproj", "{C69C243C-E40C-46CC-8599-DC7CC39F4C6F}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "workspacer.Tests", "src\workspacer.Tests\workspacer.Tests.csproj", "{FB339CF1-E0A7-4BE5-80DA-0CC095A54DE6}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "workspacer.Tests", "src\workspacer.Tests\workspacer.Tests.csproj", "{36884A7D-B921-4F6D-A324-239D79AF38F2}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "workspacer.Shared.Tests", "src\workspacer.Shared.Tests\workspacer.Shared.Tests.csproj", "{0B147785-39FE-4E5A-B205-60451444D885}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "workspacer.Bar.Tests", "src\workspacer.Bar.Tests\workspacer.Bar.Tests.csproj", "{6E6E87EB-BD73-471E-918D-C305A2CF0910}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "workspacer.Bar.Tests", "src\workspacer.Bar.Tests\workspacer.Bar.Tests.csproj", "{E07E0126-58F9-40D1-84D2-76449A902A85}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{3801228F-86AF-4076-A559-85FDE0B95200}"
```
Comment (josteink) [inline workspacer.sln:34]: Adding this solution folder is not really needed.

**Comment by josteink** [comment]
Looking over the current code, it looks pretty good.

There's some left-overs from the previous test-code changes causing the diff to be "noisier" than it needs to be, and in some places not entirely cleaned up. There's also the core business-logic of this module which might benefit from some minor testing.

Apart from that, I think this looks good, and once those things are sorted out, I'll be happy to green-light and merge this PR :+1:

File: src/workspacer/workspacer.csproj (hunk 4 above)
Comment (josteink) [reply src/workspacer/workspacer.csproj:32]: Ok. Fair enough. Let's keep it :)

File: src/workspacer.Native/Native/Win32/Win32.Long.cs (hunk 1 above)
Comment (dalyIsaac) [reply src/workspacer.Native/Native/Win32/Win32.Long.cs:13]: Browsing through docs.microsoft.com, the signatures of functions like [`SetWindowLongPtr`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowlongptra) take a `LONG_PTR`. 

[`LONG_PTR`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/bbcfb0af-8349-4e98-ad26-957e1363f714) is (logically) a pointer for the long type. 

[`LONG`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/29d44d70-382f-4998-9d76-8a1fe93e445c) is a 32-bit **signed** integer.

In .NET, [`long`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types) is a signed 64-bit integer. I changed it to be a _bit_ more correct (in my opinion), by allowing negative values (which are pretty unlikely, though I am no expert in this sort of Windows programming).

However, if you think it'd better to keep it as `ulong`, I can change it back.

File: src/workspacer.Native/Native/Win32/Win32.Long.cs (hunk 2 above)
Comment (dalyIsaac) [reply src/workspacer.Native/Native/Win32/Win32.Long.cs:40]: Browsing through docs.microsoft.com, the signatures of functions like [`SetWindowLongPtr`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowlongptra) take a `LONG_PTR`. 

[`LONG_PTR`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/bbcfb0af-8349-4e98-ad26-957e1363f714) is (logically) a pointer for the long type. 

[`LONG`](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/29d44d70-382f-4998-9d76-8a1fe93e445c) is a 32-bit **signed** integer.

In .NET, [`long`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types) is a signed 64-bit integer. I changed it to be a _bit_ more correct (in my opinion), by allowing negative values (which are pretty unlikely, though I am no expert in this sort of Windows programming).

However, if you think it'd better to keep it as `ulong`, I can change it back.

File: src/workspacer.Shared.Tests/workspacer.Shared.Tests.csproj [hunk 8]
```diff
@@ -1,7 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
-  <ItemGroup>
-    <ProjectReference Include="..\workspacer.Shared\workspacer.Shared.csproj" />
-  </ItemGroup>
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (dalyIsaac) [reply src/workspacer.Shared.Tests/workspacer.Shared.Tests.csproj:1]: Fixed in 070577a7f7b93f7a422950b420a6cec842cded18

File: src/workspacer.Tests/workspacer.Tests.csproj (hunk 3 above)
Comment (dalyIsaac) [reply src/workspacer.Tests/workspacer.Tests.csproj]: Fixed in 070577a7f7b93f7a422950b420a6cec842cded18

File: src/workspacer.TitleBar/TitleBarPlugin.cs (hunk 6 above)
Comment (dalyIsaac) [reply src/workspacer.TitleBar/TitleBarPlugin.cs:33]: I added unit tests for updating the `Win32.WS style` in 5d13fc11c2849e44e925cde08361aebb3636248c.

That commit also extracts extracts the manipulation of the `style` variable to its own static method.

File: workspacer.sln (hunk 7 above)
Comment (dalyIsaac) [reply workspacer.sln]: Fixed in 87c6f59baf4d711eca36ceb6ad7b1f3fcb73aa4a

File: src/workspacer.Native/Native/Win32/Win32.Long.cs (hunk 1 above)
Comment (josteink) [reply src/workspacer.Native/Native/Win32/Win32.Long.cs:13]: Given the currently used values I don’t think we gain much either way. But no biggie 🙂

File: src/workspacer.TitleBar/TitleBarPlugin.cs (hunk 6 above)
Comment (josteink) [reply src/workspacer.TitleBar/TitleBarPlugin.cs:33]: LGTM!

**Review by josteink — APPROVED** [review]

**Comment by josteink** [comment]
Approved and updated. Sorry about the slow response!

**Pull request merged**